Extension/src/LanguageServer/Providers/inlayHintProvider.ts

```diff
@@ -0,0 +1,112 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
+ * See 'LICENSE' in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+import * as vscode from 'vscode';
+import { DefaultClient, openFileVersions } from '../client';
+import { Position, RequestType } from 'vscode-languageclient';
+
+interface GetInlayHintsParams {
+    uri: string;
+}
+
+enum CppInlayHintKind {
+    Type = 0,
+    Parameter = 1,
+}
+
+interface CppInlayHint {
+    position: Position;
+    label: string;
+    kind: CppInlayHintKind;
+}
+
+interface GetInlayHintsResult {
+    fileVersion: number;
+    canceled: boolean;
+    inlayHints: CppInlayHint[];
+}
+
+// TODO: either use separate lists or filter
+type InlayHintsCacheEntry = { FileVersion: number, 
+    TypeHints: vscode.InlayHint[], ParameterHints: vscode.InlayHint[] }
+
+const GetInlayHintsRequest: RequestType<GetInlayHintsParams, GetInlayHintsResult, void, void> =
+    new RequestType<GetInlayHintsParams, GetInlayHintsResult, void, void>('cpptools/getInlayHints');
+
+export class InlayHintsProvider implements vscode.InlayHintsProvider {
+    private client: DefaultClient;
+    public onDidChangeInlayHintsEvent = new vscode.EventEmitter<void>();
+    onDidChangeInlayHints?: vscode.Event<void>;
+    private cache: Map<string, InlayHintsCacheEntry> = new Map<string, InlayHintsCacheEntry>();
+
+    constructor(client: DefaultClient) {
+        this.client = client;
+        this.onDidChangeInlayHints = this.onDidChangeInlayHintsEvent.event;
+    }
+
+    public async provideInlayHints(document: vscode.TextDocument, range: vscode.Range, token: vscode.CancellationToken)
+        : Promise<vscode.InlayHint[] | undefined> {
+        await this.client.awaitUntilLanguageClientReady();
+        const uriString: string = document.uri.toString();
+
+        // Get results from cache if available.
+        const cacheEntry: InlayHintsCacheEntry | undefined = this.cache.get(uriString);
+        if (cacheEntry && cacheEntry.FileVersion === document.version) {
+            return this.getHintsBasedOnSettings(cacheEntry);
+        }
+
+        // Get new results from the language server
+        const params: GetInlayHintsParams = { uri: uriString };
+        const inlayHintsResult: GetInlayHintsResult = await this.client.languageClient.sendRequest(GetInlayHintsRequest, params, token);
+        if (inlayHintsResult.canceled || (inlayHintsResult.fileVersion !== openFileVersions.get(uriString))) {
+            throw new vscode.CancellationError();

```

Comment:
Looking at code in the vscode repo, it's unclear to me whether `CancellationError` would be handled for `provideInlayHints`.  `CancellationError` was added specifically for semantic tokens, and didn't work properly for me with `FoldingRangeProvider`.  https://github.com/microsoft/vscode/issues/125799  .   In that issue, aeschli mentioned:

> CancellationError is currently only handled for semantic token provider

It might be necessary to return undefined and possibly trigger `onDidChangeInlayHints` if returning undefined does not result in a new request.  (You'll probably get a new request if VS Code cancelled the token, but might not get a new request due to a file version change.  You may need to check.)